### PR TITLE
fix: fix headers not completing when call is terminated

### PR DESCRIPTION
**Pull request by c-lucera-pvotal**
This PR attempt to solve the issue with headers not being correctly completed after a call is closed as pointed out in https://github.com/grpc/grpc-dart/issues/727

To fix the issue I have added in the `_terminate()` method a check to complete with errors any completer still going (I found `_headers` and `_trailers`)

### CODE REVIEW

**Comment by r-durao-pvotal** [comment]
@kevmoo who would be the right person to take a look at this PR? Thanks in advance!

**Comment by mraleph** [comment]
Please add a regression test and then we can land it.

**Comment by c-lucera-pvotal** [comment]
@mraleph 
Added test and fixed CI checks, both for CHANGELOG and analyze

**Comment by c-lucera-pvotal** [comment]
@mraleph 
Added the requested test and fixed CI issues like Changelog notes.

However I had to change the previous logic from throwing an error
```dart
    if (!_headers.isCompleted) {
      _headers.complete({});
    }
```

To completing with an empty value
```dart
    if (!_headers.isCompleted) {
      _headers.completeError(GrpcError.unimplemented('Request terminated before headers'));
    }
```

The reason about the change was that the error throwing was really distruptive for the current implementations, instead of receving a error only on active listeners (I.e whoever was actively looking at that completer) all the cancelled calls received 3 throws, one for the cancellation, one for the headers not completed and one for the trailers not completed, this meant that all the tests in this project were failing due to the not catched exceptions, and at the same time, anyone cancelling the call had to deal with 3 different throws, I hope this change make sense to you.

**Comment by c-lucera-pvotal** [comment]
@mraleph saw some failing tests, but those seems unrelated to my changes, any hint about those?

**Comment by mosuem** [comment]
These keepalive tests can be flaky, but this doesn't look that way. Let me check.

**Comment by c-lucera-pvotal** [comment]
It looks like that the tearDown somehow is not awaiting the closing before the next setUp is running, maybe they are running in parallel?

in [line 52 of keep_alive_test.dart,](https://github.com/grpc/grpc-dart/blob/c18e185bb09d94a6760e4c43c9bf8c05c0f0e4c9/test/keepalive_test.dart#L52C5-L52C58) if you use a different port for each serve you will see the tests not failing anymore

```dart
//before tests setup
int port = 8081;

//Line 52
 await server.serve(address: 'localhost', port: port++);
 ```
 
 Want me to add this?

**Comment by c-lucera-pvotal** [comment]
Another thing I discovered, if I run only keep alive tests it works out of the box, so probably tests are running in parallel with other tests using the port 8081, another fix is to use the port 8282 on line 52

```dart
/Line 52
 await server.serve(address: 'localhost', port: 8082);
 ```

**Comment by c-lucera-pvotal** [comment]
Basically those two tests
https://github.com/grpc/grpc-dart/blob/master/test/keepalive_test.dart
https://github.com/grpc/grpc-dart/blob/master/test/server_cancellation_test.dart

uses port 8081, this is not an issue when run one by one, but running the whole suite with `flutter test` causes he keep_alive failing because the server cancellation test run and find the port occupied, I can fix this in my pr by changing the port if you want

**Comment by c-lucera-pvotal** [comment]
Pushed the fix for `keepalive_test.dart` let me know if you prefer to revert it and solve in a different way

File: test/keepalive_test.dart [hunk 1]
```diff
@@ -49,7 +49,7 @@ void main() {
       services: [FakeEchoService()],
       keepAliveOptions: serverOptions,
     );
-    await server.serve(address: 'localhost', port: 8081);
+    await server.serve(address: 'localhost', port: 8082);
```
Comment (kevmoo) [inline test/keepalive_test.dart:52]: What's w/ this change? Necessary?

Comment (mosuem) [reply test/keepalive_test.dart:52]: This fixes the flaky keepalive tests.

Comment (c-lucera-pvotal) [reply test/keepalive_test.dart:52]: I think 0 is fine, but you need some consistency along tests, I can pass 0 in the failing one, but you should consider another PR to fix all the random port passed around (unless they are really needed) 
image

Testing with 0, if the tests passes I will push the change for keepalive_test

Comment (c-lucera-pvotal) [reply test/keepalive_test.dart]: @mraleph pushed `keepalive_test.dart` with port 0

Comment (mraleph) [reply test/keepalive_test.dart]: Thanks! We should definitely change all the tests which don't need a fixed port to use `0` as a port. This kinda slipped under the radar in reviews. We'll do it in another PR.

**Review by mraleph — APPROVED** [review]

**Pull request merged**